# PROJETO 09

## Importanto o dataset

In [1]:
import pandas as pd
import seaborn as sns

sns.set_style("whitegrid")

df = pd.read_csv('diabetes.csv')

Agora vamos estudar um pouco sobre os dados existentes no dataset de diabetes.

## Estudando os dados

In [2]:
df.describe()

Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
count   768.000000  768.000000     768.000000     768.000000  768.000000   
mean      3.845052  120.894531      69.105469      20.536458   79.799479   
std       3.369578   31.972618      19.355807      15.952218  115.244002   
min       0.000000    0.000000       0.000000       0.000000    0.000000   
25%       1.000000   99.000000      62.000000       0.000000    0.000000   
50%       3.000000  117.000000      72.000000      23.000000   30.500000   
75%       6.000000  140.250000      80.000000      32.000000  127.250000   
max      17.000000  199.000000     122.000000      99.000000  846.000000   

              BMI  DiabetesPedigreeFunction         Age     Outcome  
count  768.000000                768.000000  768.000000  768.000000  
mean    31.992578                  0.471876   33.240885    0.348958  
std      7.884160                  0.331329   11.760232    0.476951  
min      0.000000                  0.078000   21.000000    0.000000  
25%     27.300000                  0.243750   24.000000    0.000000  
50%     32.000000                  0.372500   29.000000    0.000000  
75%     36.600000                  0.626250   41.000000    1.000000  
max     67.100000                  2.420000   81.000000    1.000000

Com essa breve analise podemos observar que alguns valores, como pressão sanguinea e IMC, são iguais a 0, e esse resultado é impossível na vida real, por ser um valor que não teria como obter em uma pessoa. Então esse valor igual a 0 poderia afetar a presição dos modelos utilizados.

In [3]:
# https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.cross_val_score.html
# https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.cross_val_score.html
all_values = df.count()
missing_values = df[(df['BMI'] == 0) | (df['SkinThickness'] == 0) | (df['BloodPressure'] == 0)].count()
not_missing_values = df[(df['BMI'] != 0) & (df['SkinThickness'] != 0) & (df['BloodPressure'] != 0)].count()

print(f'Temos no total {all_values["Outcome"]} linhas e removendo as que possuem valores iguais a zero teriamos: {not_missing_values["Outcome"]}')

Temos no total 768 linhas e removendo as que possuem valores iguais a zero teriamos: 537


Temos que ao remover todos os valores iguais a 0 ficarioamos somente com 537 linhas que originalmente era de 749, para possivelmente aumentar a precisão dos modelos que serão utilizados iremos configurar essas linhas que possuem valores impossíveis (iguais a 0).

## Formatando os valores iguais a 0

A formatação será simples, basicamente iremos pegar a média desses valores que não podem valer 0 e trocaremos o 0 pela média do valor (essa média será calculada a partir de todas as linhas com o mesmo "Outcome" da linha atual) 

In [7]:
df['BMI'].replace(
    to_replace = df['BMI'][(df['BMI'] == 0) & (df['Outcome'] == 0)],
    value = df[(df['BMI'] != 0) & (df['Outcome'] == 0)]['BMI'].mean(),
    inplace = True
)

df['BMI'].replace(
    to_replace = df['BMI'][(df['BMI'] == 0) & (df['Outcome'] == 1)],
    value = df[(df['BMI'] != 0) & (df['Outcome'] == 1)]['BMI'].mean(),
    inplace = True
)

df['SkinThickness'].replace(
    to_replace = df['SkinThickness'][(df['SkinThickness'] == 0) & (df['Outcome'] == 0)],
    value = df[(df['SkinThickness'] != 0) & (df['Outcome'] == 0)]['SkinThickness'].mean(),
    inplace = True
)

df['SkinThickness'].replace(
    to_replace = df['SkinThickness'][(df['SkinThickness'] == 0) & (df['Outcome'] == 1)],
    value = df[(df['SkinThickness'] != 0) & (df['Outcome'] == 1)]['SkinThickness'].mean(),
    inplace = True
)

df['BloodPressure'].replace(
    to_replace = df['BloodPressure'][(df['BloodPressure'] == 0) & (df['Outcome'] == 0)],
    value = df[(df['BloodPressure'] != 0) & (df['Outcome'] == 0)]['BloodPressure'].mean(),
    inplace = True
)

df['BloodPressure'].replace(
    to_replace = df['BloodPressure'][(df['BloodPressure'] == 0) & (df['Outcome'] == 1)],
    value = df[(df['BloodPressure'] != 0) & (df['Outcome'] == 1)]['BloodPressure'].mean(),
    inplace = True
)

df.describe()

ValueError: Series.replace cannot use dict-like to_replace and non-None value

In [ ]:
# df[df['Outcome'] == 0]['BMI'].replace(
#     to_replace = 0,
#     value = df[(df['BMI'] != 0) & (df['Outcome'] == 0)]['BMI'].mean(),
#     inplace = True
# )

In [ ]:
# # print(df['BMI'].mean())

# # [(df['BMI'] == 0) & (df['Outcome'] == 0)]
# # df['BMI'].replace(
# #     to_replace = 0,
# #     value = df[(df['BMI'] != 0) & (df['Outcome'] == 0)]['BMI'].mean(axis=0),
# #     inplace = True
# # )

# print(df['BMI'].mean())
# df['BMI'].replace(
#     to_replace = df[(df['BMI'] == 0) & (df['Outcome'] == 0)],
#     value = df[(df['BMI'] != 0) & (df['Outcome'] == 0)]['BMI'].mean(),
#     inplace = True
# )

# # print(df['BMI'].mean())
# df['BMI'].describe()